src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp

```diff
@@ -0,0 +1,998 @@
+ *     \right),
+ * \f}
+ *
+ *
+ * Finally, limiting is performed as follows:
+ *
+ * - limit \f$c_{2,2,2}\f$, if not changed, stop
+ * - limit \f$c_{2,2,1}\f$, \f$c_{2,1,2}\f$, and \f$c_{1,2,2}\f$, if all not
+ * changed, stop
+ * - limit \f$c_{2,2,0}\f$, \f$c_{2,0,2}\f$, and \f$c_{0,2,2}\f$, if all not
+ * changed, stop
+ * - limit \f$c_{2,1,1}\f$, \f$c_{1,2,1}\f$, and \f$c_{1,1,2}\f$, if all not
+ * changed, stop
+ * - limit \f$c_{2,1,0}\f$, \f$c_{2,0,1}\f$, \f$c_{1,2,0}\f$, \f$c_{1,0,2}\f$,
+ * \f$c_{0,1,2}\f$, and \f$c_{0,2,1}\f$, if all not changed, stop
+ * - limit \f$c_{2,0,0}\f$, \f$c_{0,2,0}\f$, and \f$c_{0,0,2}\f$, if all not
+ * changed, stop
+ * - limit \f$c_{1,1,1}\f$, if not changed, stop
+ * - limit \f$c_{1,1,0}\f$, \f$c_{1,0,1}\f$, and \f$c_{0,1,1}\f$, if all not
+ * changed, stop
+ * - limit \f$c_{1,0,0}\f$, \f$c_{0,1,0}\f$, and \f$c_{0,0,1}\f$, if all not
+ * changed, stop
+ *
+ * The 1d and 2d implementations are straightforward restrictions of the
+ * described algorithm.
+ *
+ * #### Limitations:
+ *
+ * - We are recomputing the spectral coefficients after sending the coefficients
+ *   to the neighbor. We should be able to avoid this either by storing the
+ *   coefficients in the DataBox or by allowing the limiters' `packaged_data`
+ *   function to return an object to be passed as an additional argument to the
+ *   `operator()` (still would need to be stored in the DataBox).
+ * - We cannot handle the case where neighbors have more/fewer
+ *   coefficients. In the case that the neighbor has more coefficients we can
+ *   just ignore the higher coefficients. In the case that the neighbor has
+ *   fewer coefficients we have a few choices that will probably depend on which
+ *   specific HL is used.
+ * - Having different number of collocation points in different directions is
+ *   not supported. However, it is straightforward to handle this case. The
+ *   outer most loop is in the direction with the most collocation points, while
+ *   the inner most loop is over the direction with the fewest collocation
+ *   points. The highest to lowest coefficients can then be limited
+ *   appropriately again.
+ * - h-refinement is not supported, but is straightforward to add since the
+ *   differences in coefficients in Eq.\f$\ref{eq:krivodonova 3d minmod}\f$ will
+ *   just be multiplied by non-unity factors.
+ */
+template <size_t VolumeDim, typename... Tags>
+class Hierarchical<VolumeDim, tmpl::list<Tags...>> {
+ public:
+  /*!
+   * \brief The type of hierarchical limiter.
+   *
+   * Currently only the Krivodonova limiter is implemented.
+   */
+  struct Type {
+    using type = HierarchicalType;
+    static constexpr OptionString help = {"Type of hierarchical limiter"};
+  };
+
+  /*!
+   * \brief The \f$\alpha_i\f$ values in the Krivodonova algorithm.
+   *
+   * The default size is the number of allowed Legendre polynomials
+   * (`Spectral::maximum_number_of_points<Spectral::Basis::Legendre`)
```

Comment:
- "is the maximum number of allowed Legendre basis polynomials"
- missing `>` in code snippet